### ISSUE TITLE
Shared Cache not honoring group setting

### ISSUE THREAD

**Issue by wdixon** [iterative/dvc#4467]
## Bug Report

Tried to configure a dvc project with a shared cache (dvc.config entry shown below), such that multiple people belonging to the same unix group can share the cache/storage on a shared server.

Using a .dvc/cache.local configuration with:
[cache]
    type = "hardlink,symlink,copy"
    shared = group
    dir = /path/to/shared/cache_dir

However, dvc checkout --relink ends up resetting the permissions on the link files to read only, which ends up altering the permissions in the shared cache, since the hardlink shares the same inode).  When I am the owner of the files in the shared cache, the checkout succeeds; however, after resetting the cache dir permissions back to 775/664, a different user in the same group attempts a checkout and gets exception:

ERROR: unexpected error - [Errno 1] Operation not permitted: '/path/to/shared/cache_dir/17/47ffec5ae27d780ce4be074e3cc9e1.dir'

Looking at the dvc code - it appears to be attempting to chmod the file - (which fails as the user is not the owner of the cache file, or root).  There also seems to be some SHARED_MODE_MAP that isn't referenced - that may have the proper file/dir permissions to compare against?

It would seem that I have either not configured the shared cache correctly?  Or the code is not recognizing the shared cache condition, or something isn't completely implemented to support the shared cache? 

### Please provide information about your setup

**Output of `dvc version`:**

```console
$ dvc version
DVC version: 1.6.0 (conda)
---------------------------------
Platform: Python 3.8.5 on Linux-4.15.0-64-generic-x86_64-with-glibc2.10
Supports: azure, gdrive, gs, hdfs, http, https, s3, ssh, oss
Cache types: hardlink, symlink
Repo: dvc, git
```

**Additional Information (if any):**

Shared cache before checkout:
-rw-rw-r--  2 user mygroup   66203 Aug 25 08:33 feb857a8c2b62b7fac9673aa99a22e
-rw-rw-r--  2 user mygroup 1541174 Aug 25 08:34 ffd34c0b1b278750f56b27ed4c587f

Shared cache after checkout
-r--r--r--  2 user mygroup   66203 Aug 25 08:33 feb857a8c2b62b7fac9673aa99a22e
-r--r--r--  2 user mygroup 1541174 Aug 25 08:34 ffd34c0b1b278750f56b27ed4c587f

**Comment by efiop** [comment]
Hi @wdixon !

DVC expects cache files to be protected (aka with 0444 permissions) and when it stumbles upon a cache file that has incorrect permissions, it tries to verify it and chmod so it doesn't have to do that in the future. Looks like that might be a bug on our side in hardlink case, need to take a closer look...

**Comment by efiop** [comment]
Unable to reproduce right away. Do you know if maybe someone is using an old dvc version (pre 1.0) with that shared cache? Or maybe the cache has been there since those times?

**Comment by wdixon** [comment]
> Hi @wdixon !
> 
> DVC expects cache files to be protected (aka with 0444 permissions) and when it stumbles upon a cache file that has incorrect permissions, it tries to verify it and chmod so it doesn't have to do that in the future. Looks like that might be a bug on our side in hardlink case, need to take a closer look...


I used this link from the documentation that directed me to set the permissions to 0664 for the cache files?  https://dvc.org/doc/use-cases/shared-development-server

But your comet leads me to believe that perhaps the documentation is wrong?

**Comment by wdixon** [comment]
> Unable to reproduce right away. Do you know if maybe someone is using an old dvc version (pre 1.0) with that shared cache? Or maybe the cache has been there since those times?

We were using 1.6.0 - but perhaps this is a documentation issue?  I will repeat the test with the cache files set to 0444

**Comment by wdixon** [comment]
It looks like if the cache is set to 775/444 - its working...  If I followed the instructions ( https://dvc.org/doc/use-cases/shared-development-server ) which has us modify the permissions to 775/664.  If am the first person to interact with the cache (as the owner of all the files), dvc does change the file permission to 444; however, if someone else first interacts with the cache with the permissions left at 664 - an exception is thrown as the chmod fails (with the other person not being the file owner).

So I think if the documentation were updated to 775/444 - it would have worked as expected.
